bt_navigator:
  ros__parameters:
    global_frame: map
    robot_base_frame: base_link
    odom_topic: /global/odom
    plugin_lib_names:
    - nav2_compute_path_to_pose_action_bt_node
    - nav2_compute_path_through_poses_action_bt_node
    - nav2_follow_path_action_bt_node
    - nav2_spin_action_bt_node
    - nav2_wait_action_bt_node
    - nav2_back_up_action_bt_node
    - nav2_clear_costmap_service_bt_node
    - nav2_goal_reached_condition_bt_node
    - nav2_goal_updated_condition_bt_node
    - nav2_globally_updated_goal_condition_bt_node
    - nav2_is_path_valid_condition_bt_node
    - nav2_rate_controller_bt_node
    - nav2_recovery_node_bt_node
    - nav2_pipeline_sequence_bt_node
    - nav2_round_robin_node_bt_node
    - nav2_remove_passed_goals_action_bt_node

controller_server:
  ros__parameters:
    controller_frequency: 20.0
    failure_tolerance: 1.0
    min_x_velocity_threshold: 0.001
    min_y_velocity_threshold: 0.5
    min_theta_velocity_threshold: 0.001
    progress_checker_plugin: "progress_checker"
    goal_checker_plugins: ["general_goal_checker"]
    controller_plugins: ["GracefulController"]

    # Progress checker parameters
    progress_checker:
      plugin: "nav2_controller::SimpleProgressChecker"
      required_movement_radius: 0.5
      movement_time_allowance: 10.0
    # Goal checker parameters
    general_goal_checker:
      plugin: "nav2_controller::SimpleGoalChecker"
      xy_goal_tolerance: 2.0
      yaw_goal_tolerance: 0.5
      stateful: True
    # Controller parameters
    GracefulController:
      plugin: "graceful_controller/GracefulControllerROS"
      min_vel_x: 0.1
      max_vel_x: 0.5
      max_vel_theta: 2.0
      min_in_place_vel_theta: 0.6
      acc_lim_x: 1.0
      acc_lim_theta: 1.0
      max_lookahead: 1.25
      initial_rotate_tolerance: 0.25
      prefer_final_rotation: true
      yaw_filter_tolerance: 0.785
      publish_collision_points: true

local_costmap:
  local_costmap:
    ros__parameters:
      update_frequency: 5.0
      publish_frequency: 2.0
      global_frame: odom
      robot_base_frame: base_link
      rolling_window: true
      width: 6
      height: 6
      resolution: 0.1
      robot_radius: 0.25
      footprint_padding: 0.025
      plugins: ["obstacle_layer", "inflation_layer"]
      obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        publish_observations: True
        enabled: True
        publish_voxel_map: True
        origin_z: 0.0
        z_resolution: 0.125
        z_voxels: 16
        min_obstacle_height: 0.0
        max_obstacle_height: 2.0
        obstacle_min_range: 0.0
        obstacle_max_range: 5.0
        raytrace_min_range: 0.0
        raytrace_max_range: 6.0
        mark_threshold: 0
        observation_sources: livox
        livox:
          topic: /livox/lidar
          max_obstacle_height: 2.0
          clearing: True
          marking: True
          data_type: "PointCloud2"
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        cost_scaling_factor: 10.0
        inflation_radius: 1.0
      always_send_full_costmap: True

global_costmap:
  global_costmap:
    ros__parameters:
      update_frequency: 1.0
      publish_frequency: 1.0
      global_frame: map
      robot_base_frame: base_link
      rolling_window: true
      robot_radius: 0.25
      footprint_padding: 0.025
      resolution: 0.1
      plugins: ["obstacle_layer", "inflation_layer"]
      obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        enabled: True
        observation_sources: livox
        livox:
          topic: /livox/lidar
          max_obstacle_height: 2.0
          clearing: True
          marking: True
          data_type: "PointCloud2"
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        cost_scaling_factor: 10.0
        inflation_radius: 1.0
      always_send_full_costmap: True

planner_server:
  ros__parameters:
    expected_planner_frequency: 20.0
    planner_plugins: ["GridBased"]
    GridBased:
      plugin: "nav2_navfn_planner/NavfnPlanner"
      tolerance: 0.5
      use_astar: false
      allow_unknown: true

behavior_server:
  ros__parameters:
    costmap_topic: local_costmap/costmap_raw
    footprint_topic: local_costmap/published_footprint
    cycle_frequency: 10.0
    behavior_plugins: ["spin", "backup", "wait"]
    spin:
      plugin: "nav2_behaviors/Spin"
    backup:
      plugin: "nav2_behaviors/BackUp"
    wait:
      plugin: "nav2_behaviors/Wait"
    global_frame: odom
    robot_base_frame: base_link
    transform_timeout: 0.1
    # Parameters for spin recovery
    simulate_ahead_time: 2.0
    max_rotational_vel: 2.8
    min_rotational_vel: 0.6
    rotational_acc_lim: 1.0
